darshan-util/pydarshan/darshan/experimental/plots/data_access_by_filesystem.py

```diff
@@ -0,0 +1,524 @@
+    A tuple of form: `(read_groups, write_groups)` where
+    each tuple element is a `pd.Series` object adjusted to
+    contain "0" values for inactive filesystem roots (for
+    plotting purposes).
+    """
+    read_groups = empty_series_handler(series=read_groups,
+                                       filesystem_roots=filesystem_roots)
+    write_groups = empty_series_handler(series=write_groups,
+                                       filesystem_roots=filesystem_roots)
+    return read_groups, write_groups
+
+def empty_series_handler(series, filesystem_roots: List[str]):
+    """
+    Parameters
+    ----------
+
+    series: a ``pd.Series`` object
+
+    filesystem_roots: a list of strings containing unique filesystem root paths
+
+    Returns
+    -------
+    A new ``Series`` with any missing filesystem_root indices
+    filled in along with count values of 0 (for plotting purposes).
+    """
+    return series.reindex(filesystem_roots, fill_value=0).astype(np.float64)
+
+
+def rec_to_rw_counter_dfs_with_cols(report: Any,
+                                    file_id_dict: Dict[int, str],
+                                    mod: str = 'POSIX'):
+    """
+    Filter a DarshanReport into two "counters" dataframes,
+    each with read/write activity in each row (at least 1
+    byte read or written per row) and two extra columns
+    for filesystem root path and filename data.
+
+    Parameters
+    ----------
+
+    report: a darshan.DarshanReport()
+
+    file_id_dict: a dictionary mapping integer file hash values
+                  to string values corresponding to their respective
+                  paths
+
+    mod: a string indicating the darshan module to use for parsing
+         (default: ``POSIX``)
+
+
+    Returns
+    -------
+    tuple of form: (df_reads, df_writes)
+    """
+    # filter the DarshanReport into two "counters" dataframes
+    # with read/write activity in each row (at least 1
+    # byte read or written per row)
+    df_reads, df_writes = rec_to_rw_counter_dfs(report=report,
+                                                mod=mod)
+
+    # add columns with filepaths and filesystem root
+    # paths for each event
+    df_reads, df_writes = add_filesystem_cols(df_reads=df_reads,
+                                              df_writes=df_writes,
+                                              file_id_dict=file_id_dict)
+    return df_reads, df_writes
+
+def rec_to_rw_counter_dfs(report: Any,
+                          mod: str = 'POSIX'):
+    """
+    Filter a DarshanReport into two "counters" dataframes,
+    each with read/write activity in each row (at least 1
+    byte read or written per row).
+
+    Parameters
+    ----------
+
+    report: a darshan.DarshanReport()
+
+    mod: a string indicating the darshan module to use for parsing
+         (default: ``POSIX``)
+
+    Returns
+    -------
+    tuple of form: (df_reads, df_writes)
+    """
+    report.mod_read_all_records(mod, dtype='pandas')
```

Comment:
Thanks for putting this together, looks to be coming along nicely!

On the 2 issues you mention:

1.  If it's not too much trouble, I'd try to keep the labels representing bytes/files to some fixed offset from the end of the corresponding bar (i.e., not at the far right bound of the plot and not some distance proportional to the size of the bar). Just a constant-width space from the end of the bar so you don't have to scan too much to find the value.
2. I agree that we should definitely use a shared x-axis to allow easy comparison across file systems. If anything, we could consider logarithmic axes at some point (as we've discussed in other plots) if there's a wide disparity in values that makes things visually unpleasant.

One question: what do the numbers to the far left on the first 3 rows correspond to? The last row is a file system mount point (/global), like I'd expect, but the others appear more like the hash values we use to represent individual files.

Reply:
We should definitely spend some time benchmarking things at some point to see if we can alleviate some performance bottlenecks for larger log files. 2 minutes is just way too long to wait on a report. I think @jakobluettgau has thought about this some and had some ideas, including some ways we could help alleviate this from the darshan-util library itself. Not sure where it fits on the priority list, but happy to help figure it out.

Reply:
> One question: what do the numbers to the far left on the first 3 rows correspond to? The last row is a file system mount point (/global), like I'd expect, but the others appear more like the hash values we use to represent individual files.

At the moment my guess is as good as yours. I'm assuming there is code somewhere that couldn't find a file system name to retrieve from the hash/path dictionary, so it just kept the hash instead, but I can't confirm that at the moment. I'll have to trace it back when I have time.

Reply:
I thought I already pinged Shane and Phil about this below

Reply:
> We should definitely spend some time benchmarking things at some point to see if we can alleviate some performance bottlenecks for larger log files. 2 minutes is just way too long to wait on a report. I think @jakobluettgau has thought about this some and had some ideas, including some ways we could help alleviate this from the darshan-util library itself. Not sure where it fits on the priority list, but happy to help figure it out.

I have proposed some performance improvements here (see [comment](https://github.com/darshan-hpc/darshan/pull/397#issuecomment-889486713)), I just have to check them more thoroughly for correctness. I also took a look at `darshan.backend.cffi_backend.log_get_generic_record()`, and there are some optimizations I could do there to further increase performance there since it is the second-largest bottleneck. I would have to write tests to feel comfortable pushing those changes, but one of my local branches is running the same log in ~1.7 seconds (instead of ~120).  

I don't know that we have set any specific thresholds, but is there an upper limit on how long we want it to take to generate the entire job summary report (for a given log size)? If I had a number in mind it would be a lot easier to justify looking into performance bottlenecks.

Reply:
If performance stuff is getting serious, I would start using `asv` to probe suggested changes across a range of log file sizes: https://github.com/darshan-hpc/darshan/issues/426

Reply:
Sorry, probably would have been less confusing for me to comment further down on a couple of these issues after reading further...

Looks like @tylerjereddy opened an issue (#426) for us to track performance concerns going forward. I think there are a few broader components to the overhead that would be good to have a better grasp on:

1. darshan-util library raw performance for pulling records from a log file
2. overhead caused by our Python bindings for exposing log data to Python apps
3. overhead of analysis scripts like the new job-summary tool for processing log data and generating graphs in Python

As far as what a reasonable overhead is, it would just be good if the analysis scripts didn't take too much longer than our raw performance for reading the log file with the C library. I guess that's vague, but partly because I don't know just how slow we are in the C library for reading massive logs -- I know it already takes us O(tens of seconds) to parse some logs, but that's mostly dominated by needing to write massive amounts of text related to log file contents (to stdout or redirected to a file). I think once we have a better grasp on 1 and 2 above we could say more directly how much overhead we're adding in these new analysis scripts. The 1.7 seconds @nawtrey has gotten this down to sounds completely fine for now.

Reply:
At the moment the majority of the run time is spent upstream in `report.mod_read_all_records()`. The worst logs I've tried are taking 16-20 seconds to generate a figure, where 90-95% of that time is spent in the aforementioned function. We've been working on addressing some of the bottlenecks there, so once those changes are merged this should be generating the figure sufficiently fast. I'm not sure if that should be a requirement for merging though, since this is in the `/experimental`.

Reply:
After the discussion in [#490](https://github.com/darshan-hpc/darshan/issues/490) I think we can remove this line all together. The [following line](https://github.com/darshan-hpc/darshan/pull/397/files?authenticity_token=jBtkvZELuIzGOM9viM5VnkwaaYcoHt9A%2BT57V7qWsTzGg5RSmc2fAvLP1i9NMFSFVjbji5U1%2FrQ3EMJKruS4jw%3D%3D&file-filters%5B%5D=.py#diff-bbb29285c3073260dfbe4f3c6175265b7623c05c8eb36737469a99d20a0aa74fR189) could be changed to use `.to_df()` instead.

Reply:
I've updated this code to use `to_df()` instead of the deprecated-but-not-deprecated `dtype=` stuff. I'm not sure if this resolves the full discussion above--I'll leave that to reviewers to decide.

Reply:
For most of the available logs I'm seeing <1 second runtimes to generate a figure, where most of this time (~80%) is spent converting records to a dataframe in `to_df()`. I'll resolve this since the remaining performance concerns reside elsewhere.